File: bindings/java/src/io/quad4/rrc/Buffers.java
// SPDX-License-Identifier: 0BSD

package io.quad4.rrc;

import com.sun.jna.ptr.LongByReference;
import io.quad4.rrc.ffi.RrcLibrary;

final class Buffers {
    private static final int MAX_BUFFER = 16 * 1024 * 1024;

    private Buffers() {}

    @FunctionalInterface
    interface StringFill {
        int fill(byte[] buf, int buflen, LongByReference written);
    }

    @FunctionalInterface
    interface BytesFill {
        int fill(byte[] buf, int buflen, LongByReference written);
    }

    static String readString(StringFill fill, int initial) {
        int cap = initial;
        while (cap <= MAX_BUFFER) {
            byte[] buf = new byte[cap];
            LongByReference written = new LongByReference();
            int code = fill.fill(buf, buf.length, written);
            if (code == RrcException.TRUNCATED) {
                cap *= 2;
                continue;
            }
            Rrc.check(code);
            return Rrc.cString(buf, (int) written.getValue());
        }
        throw new RrcException(RrcException.TRUNCATED);
    }

    static byte[] writeBytes(BytesFill fill, int initial) {
        int cap = initial;
        while (cap <= MAX_BUFFER) {
            byte[] buf = new byte[cap];
            LongByReference written = new LongByReference();
            int code = fill.fill(buf, buf.length, written);
            if (code == RrcException.TRUNCATED) {
                cap *= 2;
                continue;
            }
            Rrc.check(code);
            byte[] out = new byte[(int) written.getValue()];
            System.arraycopy(buf, 0, out, 0, out.length);
            return out;
        }
        throw new RrcException(RrcException.TRUNCATED);
    }
}

File: bindings/lua/rrc/buffers.lua
-- SPDX-License-Identifier: 0BSD

local ffi = require("ffi")

local M = {}

M.MAX_BUFFER = 16 * 1024 * 1024
M.RRC_ERR_TRUNCATED = 8

function M.read_string(fill, initial)
	local cap = initial or 1024
	while cap <= M.MAX_BUFFER do
		local buf = ffi.new("char[?]", cap)
		local written = ffi.new("size_t[1]")
		local code = fill(buf, cap, written)
		if code == M.RRC_ERR_TRUNCATED then
			cap = cap * 2
		else
			return code, ffi.string(buf, written[0])
		end
	end
	return M.RRC_ERR_TRUNCATED, nil
end

function M.write_bytes(fill, initial)
	local cap = initial or 65536
	while cap <= M.MAX_BUFFER do
		local buf = ffi.new("uint8_t[?]", cap)
		local written = ffi.new("size_t[1]")
		local code = fill(buf, cap, written)
		if code == M.RRC_ERR_TRUNCATED then
			cap = cap * 2
		else
			return code, ffi.string(buf, written[0])
		end
	end
	return M.RRC_ERR_TRUNCATED, nil
end

return M
